File: backend/app/services/ai_service.py
"""
AI Service - Groq API Integration
Uses openai/gpt-oss-20b model via Groq
"""
import httpx
from typing import Dict, Any, Optional
from app.config import get_settings

settings = get_settings()


class AIService:
    def __init__(self, api_key: str):
        self.api_key = api_key
        self.base_url = "https://api.groq.com/openai/v1/chat/completions"
        # Groq models: llama-3.3-70b-versatile, llama-3.1-70b-versatile, mixtral-8x7b-32768, gemma2-9b-it
        self.model = "llama-3.3-70b-versatile"

    async def verify_attendance(self, evidence: Dict[str, Any]) -> Dict[str, Any]:
        """
        Send attendance evidence to AI for verification.

        Args:
            evidence: {
                "student_code": "HV001",
                "gps_distance": 18,
                "gps_accuracy": 12,
                "qr_valid": True,
                "check_in_time": "08:02",
                "session_start": "08:00"
            }

        Returns:
            {
                "status": "CONFIRMED" | "VERIFY" | "SUSPICIOUS",
                "confidence": 0-100,
                "reasoning": "...",
                "uncertainty": []
            }
        """
        prompt = self._build_prompt(evidence)

        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json"
        }

        payload = {
            "model": self.model,
            "messages": [
                {
                    "role": "system",
                    "content": """You are a strict attendance verification AI. You MUST follow these rules EXACTLY:

STRICT RULES:
1. GPS < 50m → CONFIRMED
2. GPS 50-100m → VERIFY
3. GPS > 100m → SUSPICIOUS

The GPS distance is the PRIMARY factor. Follow the rules above EXACTLY, no exceptions.

Return ONLY valid JSON, no markdown, no explanation:
{"status": "CONFIRMED", "confidence": 95, "reasoning": "GPS distance is within acceptable range", "uncertainty": []}"""
                },
                {
                    "role": "user",
                    "content": prompt
                }
            ],
            "temperature": 0.0,
            "response_format": {"type": "json_object"}
        }

        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.post(self.base_url, json=payload, headers=headers)

            if response.status_code != 200:
                raise Exception(f"AI API error: {response.status_code} - {response.text}")

            data = response.json()
            content = data.get("choices", [{}])[0].get("message", {}).get("content", "{}")

            import json
            result = json.loads(content)

            return {
                "status": result.get("status", "VERIFY").upper(),
                "confidence": result.get("confidence", 50),
                "reasoning": result.get("reasoning", ""),
                "uncertainty": result.get("uncertainty", [])
            }

    def _build_prompt(self, evidence: Dict[str, Any]) -> str:
        return f"""Attendance Evidence:
- Student Code: {evidence.get('student_code', 'N/A')}
- GPS Distance from classroom: {evidence.get('gps_distance', 'N/A')}m
- GPS Accuracy: {evidence.get('gps_accuracy', 'N/A')}m
- QR Token Valid: {'Yes' if evidence.get('qr_valid') else 'No'}
- Check-in Time: {evidence.get('check_in_time', 'N/A')}
- Session Start Time: {evidence.get('session_start', 'N/A')}

Analyze if all these signals are consistent with the student being physically present."""


# Singleton instance
_ai_service: Optional[AIService] = None


def get_ai_service() -> AIService:
    global _ai_service
    if _ai_service is None:
        _ai_service = AIService(api_key=settings.GROQ_API_KEY)
    return _ai_service

File: UI/lib/ai-service.ts
// lib/ai-service.ts
// AI service cho AI Attendance - Presence Verification

export type Verdict = "confirmed" | "verify" | "suspicious"

export type VerificationInput = {
  time: string        // VD: "10:03:12"
  gps: number         // VD: 12 (meters)
  deviceMatched: boolean
  tokenValid: boolean
}

export type VerificationResult = {
  verdict: Verdict
  confidence: number  // 0-100
  reason: string
}

// Groq API call
export async function verifyPresence(input: VerificationInput): Promise<VerificationResult> {
  const apiKey = process.env.GROQ_API_KEY
  console.log("🔑 Groq API Key:", apiKey ? "loaded (length=" + apiKey.length + ")" : "NOT FOUND")

  if (!apiKey) {
    console.warn("GROQ_API_KEY not set, using mock result")
    return mockVerify(input)
  }

  const prompt = `You are an attendance verification AI.

Analyze the attendance evidence below. Determine whether the evidence is consistent with the student being physically present in the classroom.

Evidence:
- GPS distance from classroom: ${input.gps}m
- Check-in timestamp: ${input.time}
- QR Token valid: ${input.tokenValid ? "Yes" : "No"}
- Device ID matched: ${input.deviceMatched ? "Yes" : "No"}

Consider ALL signals together. Think about:
- GPS distance: is ${input.gps}m reasonable? Below 50m is ideal, 50-100m needs verification, above 100m is suspicious.
- Check-in timing: consistent with normal class start patterns?
- QR validity: expired or invalid tokens suggest suspicious behavior
- Device matching: mismatched devices suggest potential proxy attendance

Return a JSON object with these fields:
{
  "verdict": "CONFIRMED" or "VERIFY" or "SUSPICIOUS",
  "confidence": 0-100,
  "reasoning": "Your analysis considering ALL signals together",
  "uncertain_evidence": "What specific evidence caused uncertainty (if any)"
}

Example reasoning patterns:
- "GPS is within acceptable range and all other signals are consistent → CONFIRMED"
- "GPS is borderline and timestamp is unusual → VERIFY for manual review"
- "GPS far from classroom AND QR invalid → SUSPICIOUS"`

  try {
    const response = await fetch(
      "https://api.groq.com/openai/v1/chat/completions",
      {
        method: "POST",
        headers: {
          "Authorization": `Bearer ${apiKey}`,
          "Content-Type": "application/json"
        },
        body: JSON.stringify({
          model :"openai/gpt-oss-20b",  // Open source model ~14B params
          messages: [
            { role: "system", content: "Bạn là AI xác minh điểm danh. Luôn trả lời đúng định dạng JSON." },
            { role: "user", content: prompt }
          ],
          temperature: 0.1,
          response_format: { type: "json_object" }
        })
      }
    )

    if (!response.ok) {
      const errorText = await response.text()
      console.error("Groq API error:", response.status, errorText)
      return mockVerify(input)
    }

    const data = await response.json()
    const content = data.choices?.[0]?.message?.content

    if (!content) {
      console.warn("No content from Groq, using mock")
      return mockVerify(input)
    }

    console.log("📦 Groq response:", content)

    // Parse JSON response
    const parsed = JSON.parse(content)
    // Normalize verdict to lowercase
    const normalizedVerdict = (parsed.verdict || "verify").toLowerCase() as Verdict
    return {
      verdict: normalizedVerdict,
      confidence: Math.min(100, Math.max(0, parsed.confidence || 50)),
      reason: parsed.reasoning || parsed.reason || "Không có lý do"
    }
  } catch (error) {
    console.error("AI verification error:", error)
    return mockVerify(input)
  }
}

// Mock fallback khi không có API key
function mockVerify(input: VerificationInput): VerificationResult {
  let verdict: Verdict = "confirmed"
  let confidence = 95
  let reasons: string[] = []

  // GPS check
  if (input.gps > 100) {
    verdict = "suspicious"
    confidence = Math.max(10, 100 - input.gps)
    reasons.push(`GPS ${input.gps}m quá xa`)
  } else if (input.gps > 50) {
    verdict = "verify"
    confidence = Math.max(50, 85 - (input.gps - 50))
    reasons.push(`GPS ${input.gps}m gần boundary`)
  } else {
    reasons.push(`GPS ${input.gps}m trong phạm vi`)
  }

  // Token check
  if (!input.tokenValid) {
    verdict = "suspicious"
    confidence = Math.min(confidence, 30)
    reasons.push("Token không hợp lệ")
  }

  // Device check
  if (!input.deviceMatched) {
    if (verdict === "suspicious") {
      confidence = Math.min(confidence, 15)
    } else {
      verdict = "verify"
      confidence = Math.min(confidence, 60)
    }
    reasons.push("Device không match")
  }

  return {
    verdict,
    confidence,
    reason: reasons.join(", ") || "Tất cả evidence đạt"
  }
}

// Export cho test runner
export { verifyPresence as default }
